HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +53,72 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
-
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
 
+    cdef:
+        int idx, r
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
+    cdef:
+        int idx, r
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+        int source, destination, case
 
    ##########
    # Your code here
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
 
-   free_N_locks(counts.shape[0], locks)
+    with nogil:
+        for r in prange(repeat, num_threads=4):
+            # Everything in here is parallel...so we have to be very careful. We can easily cause deadlock...
+            for idx in range(src.shape[0]):
+                # We need to lock here...but we can accidentally do terrible things. We need a lock on both
+                # the source and the destination...
+
+                source = src[idx]
+                destination = dest[idx]
+
+                # To avoid deadlock, always lock the smaller one first.
+                if source < destination: case = -1
+                elif source > destination: case=1
+                else: case=0
+
+                if case==-1:
+                    acquire(&locks[source])
+                    acquire(&locks[destination])
+                elif case==0:
+                    acquire(&locks[source])
+                elif case==1:
+                    acquire(&locks[destination])
+                    acquire(&locks[source])
+
+                # Do the transfer
+                if counts[src[idx]] > 0:
+                    counts[destination] += 1
```

Comment:
I think for the fine-grained case, you can just lock destination and source separately when writing to counts. It is sufficient to guarantee that the same amount of transfers happens as in single-threaded case. (Note that it isn't exactly equivalent to the serial version logic-wise because it can double-increment or double-decrement some items, but is good enough for the correctness condition of this part).